### PR TITLE
Fix `XGLMModelLanguageGenerationTest.test_batched_nan_fp16`

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

#18057 added this test to test running with fp16.

However, `from_pretrained(model_name, torch_dtype=torch.float16` seems **not able to change the dtype** for weights registered below:
https://github.com/huggingface/transformers/blob/a7bc4221c0c09857b30ac467e7de86d3f5a7c482/src/transformers/models/xglm/modeling_xglm.py#L168-L176
and `hidden_states` becomes again `float32` (because `position` is) at
https://github.com/huggingface/transformers/blob/a7bc4221c0c09857b30ac467e7de86d3f5a7c482/src/transformers/models/xglm/modeling_xglm.py#L715
and finally failed at `hidden_states = self.self_attn_layer_norm(hidden_states)` with
```bash
RuntimeError: expected scalar type Float but found Half
```

### CODE REVIEW

File: src/transformers/models/xglm/modeling_xglm.py [hunk 1]
```diff
@@ -709,8 +709,8 @@ def forward(
             # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
             encoder_attention_mask = _expand_mask(encoder_attention_mask, inputs_embeds.dtype, tgt_len=input_shape[-1])
 
-        # embed positions
-        positions = self.embed_positions(input_ids, inputs_embeds, past_key_values_length)
+        # embed positions, cast from float32 to `inputs_embeds.dtype`
+        positions = self.embed_positions(input_ids, inputs_embeds, past_key_values_length).to(inputs_embeds.dtype)
```
Comment (ydshieh) [inline src/transformers/models/xglm/modeling_xglm.py:713]: position (and `embed_positions`'s weight) are in float32 even if we load the model in float16. Need this cast to make later LayerNorm layer work.

Comment (sgugger) [reply src/transformers/models/xglm/modeling_xglm.py:713]: I think the right fix would be to make sure the weights have the correct dtype, as the embedding layer is the biggest one, so the memory save is very important.

**Review by sgugger — COMMENTED** [review]
Thanks for looking into this. While this fixes the issue, I'm not sure if it's the right fix.

**Comment by ydshieh** [comment]
@sgugger OK, let me check if I can do something for (not a real) weights defined by

```python
self.register_buffer("weights", emb_weights)

```

**Review by sgugger — APPROVED** [review]
Thanks, this feels like a much better fix!

**Pull request merged**